astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+                                                 preserve_observer_frame=preserve_observer_frame)
+
+        new_coord = self._change_observer_to(observer)
+
+        return new_coord
+
+    def with_radial_velocity_shift(self, target_shift=None, observer_shift=None):
+        """
+        Apply a velocity shift to this spectral coordinate. The shift
+        can be provided as a redshift (float value) or radial velocity
+        (`~astropy.units.Quantity` with physical type of 'speed').
+
+        Parameters
+        ----------
+        target_shift : float or `~astropy.units.Quantity`
+            Shift value to apply to current target.
+        observer_shift : float or `~astropy.units.Quantity`
+            Shift value to apply to current observer.
+
+        Returns
+        -------
+        `SpectralCoord`
+            New spectral coordinate with the target/observer velocity changed
+            to incorporate the shift.
+        """
+        if observer_shift is not None and (self.target is None or
+                                           self.observer is None):
+            raise ValueError("Both an observer and target must be defined "
+                             "before applying a velocity shift.")
+
+        for arg in [x for x in [target_shift, observer_shift] if x is not None]:
+            if isinstance(arg, u.Quantity) and \
+                    arg.unit.physical_type not in ['speed', 'dimensionless']:
+                raise u.UnitsError("Argument must have unit physical type "
+                                   "'speed' for radial velocty or "
+                                   "'dimesionless' for redshift.")
+
+        # The target or observer value is defined but is not a quantity object,
+        #  assume it's a redshift float value and convert to velocity
+
+        if target_shift is not None:
+            target_shift = u.Quantity(target_shift)
+            if target_shift.unit.physical_type == 'dimensionless':
+                target_shift = _redshift_to_velocity(target_shift)
+
+        if observer_shift is not None:
+            observer_shift = u.Quantity(observer_shift)
+            if observer_shift.unit.physical_type == 'dimensionless':
+                observer_shift = _redshift_to_velocity(observer_shift)
+
+        target_icrs = self._target.transform_to(ICRS)
+        observer_icrs = self._observer.transform_to(ICRS)
+
+        target_shift = 0 * u.km / u.s if target_shift is None else target_shift
+        observer_shift = 0 * u.km / u.s if observer_shift is None else observer_shift
+
+        pos_hat = SpectralCoord._norm_d_pos(observer_icrs, target_icrs)
+
+        target_velocity = target_icrs.velocity + target_shift * pos_hat
+        observer_velocity = observer_icrs.velocity + observer_shift * pos_hat
+
+        new_target = target_icrs.realize_frame(
+            target_icrs.cartesian.with_differentials(
+                CartesianDifferential(target_velocity.xyz))).transform_to(
```

Comment:
I think `target_velocity` is always a `CartesianDifferential`, no?

Reply:
No, it seems it is a ``CartesianRepresentation`` because above we add a differential to a representation so we get a representation. So I still need the explicit conversion to a differential.